### DISCUSSION TITLE
Problem regarding leftsubnet and mode-cfg ip

### DISCUSSION THREAD

**Discussion by slemke76** [strongswan/strongswan#1170] [General]
Hello,

I have a roadwarrior setup – a client with a dynamic ip address is connecting to the headquarter. On the client side strongswan is being used, on the server-side a Fortigate is working. It is an IKEv2 tunnel, main mode, public key authentification, with mode-cfg enabled. The fortigate assigns an ip from the range 10.255.10.62.2 – 10.255.10.62.254 with mode-cfg for the tunnel. The left side of the tunnel is the 10.255.10.0/24 subnet. My goal is to reach devices on the left subnet **AND** the ip assigned by mode-cfg to strongswan (as a backup, when the left subnet changes). I do not need access from the left-side to the right-side, only from right to left.

Here is my config:
```
conn hq2
    also=vip-base
    keyexchange=ikev2
    leftauth=pubkey
    rightsubnet=192.168.2.0/24
    leftsubnet=10.255.10.0/24
    ike=aes256-sha512-modp4096!
    ikelifetime=86400s
    esp=aes256-sha512-modp4096!
    lifetime=43200s

    leftcert=RPi_Notfall-Router_01.pem
    leftid="C=DE, ST=Example, L=Example, O=Example, CN=RPi Notfall-Router 01"

    right=hq.example.org
    rightid="C=DE, ST=Example, L=Example, O=Example, CN=hq.example.org"
    rightauth=pubkey
    auto=start

```
I have tested several scenarios:

```
leftsubnet=10.255.10.0/24
CHILD_SA hq2{3} established with SPIs c5a91f48_i be8259ce_o and TS 10.255.10.0/24 === 192.168.2.0/24
```

```
leftsubnet=10.255.62.2/32
CHILD_SA hq2{3} established with SPIs ca7408f1_i be8259cb_o and TS 10.255.62.2/32 === 192.168.2.0/24

```
```
leftsubnet=10.255.10.0/24,10.255.62.2/32
CHILD_SA hq2{2} established with SPIs cfdb0132_i be8259d2_o and TS 10.255.10.0/24 10.255.62.2/32 === 192.168.2.0/24

```
All these cases are working properly, but when the fortigate assigns with mode-cfg another ip than 10.255.62.2/32 it will not work anymore. So I need something like this for my definition

`leftsubnet=10.255.10.0/24,%dynamic
`

But this seems not to work. Even this does not work:

```
leftsubnet=%dynamic
failed to establish CHILD_SA, keeping IKE_SA

```
I guess “%dynamic” is not a valid parameter for leftsubnet, is there a way to solve this?

Thanks,
Sebastian

**Comment by tobiasbrunner** [comment]
> It is an IKEv2 tunnel, main mode, public key authentification, with mode-cfg enabled.

Just FYI, main mode and mode-cfg are IKEv1 concepts. To assign virtual IPs etc. IKEv2 uses configuration payloads/attributes during IKE_AUTH, not a separate intermediate exchange/mode.

> `leftsubnet=10.255.10.0/24,%dynamic`

I guess that's the correct setting if you want both a virtual IP and the subnet as local traffic selectors. Note that this config will result in `0.0.0.0/0` getting proposed as local traffic selector because that's basically what `%dynamic` means and the other subnet is fully contained in it and gets eliminated before sending the TSi payload. This requires a corresponding configuration and narrowing behavior on the peer. Maybe the other peer explicitly expects a subnet besides `0.0.0.0/0` (even though it could assign whatever subnets and IPs it wants to the latter - might also depend on its configuration, so you might want to check what traffic selectors are configured there).

You could try if `leftsubnet=10.255.10.0/24,10.255.62.0/24` makes a difference, but if the responder doesn't narrow the second traffic selector to the assigned virtual IP this will redirect traffic to all VIPs to your host (and it will break if the subnet for virtual IPs ever changes). Another option is explicitly configuring `leftsubnet=0.0.0.0/0`, but, again, it will depend on the responder's behavior (especially to avoid redirecting all traffic to you).

> But this seems not to work.

What exactly is the error?  Could you please post the logs (see #196)?

**Reply by slemke76** [reply]
Hello,

> Just FYI, main mode and mode-cfg are IKEv1 concepts.  [..]

Thanks to make this clear to me - I have primary used IKEv1 up to now.

I have debugged the other side (Fortigate) - as you told:

```
leftsubnet=10.255.10.0/24,%dynamic
results in this TSi payload: 
TSi_0 0:0.0.0.0-255.255.255.255:0

```
and 

```
leftsubnet=10.255.10.0/24,10.255.62.0/24
in  
TSi_0 0:10.255.10.0-10.255.10.255:0
TSi_1 0:10.255.62.0-10.255.62.255:0

```
So the responder does not narrow the selector.

Is there a way to configure stronswan in that way that it proposes the 10.255.10.0/24 net and additionally the assigned (mode-cfg) ip and not 0.0.0.0/0 ?

I will now additionally check if I have any options to configure the responder.

Thanks
Sebastian

**Reply by tobiasbrunner** [reply]
> So the responder does not narrow the selector.

It's weird that it doesn't narrow any of the two matching TS down to the assigned virtual IP. It even accepts that the client redirects **all** traffic to it (first attempt).

> Is there a way to configure stronswan in that way that it proposes the 10.255.10.0/24 net and additionally the assigned (mode-cfg) ip and not 0.0.0.0/0 ?

As I mentioned, this information is generally returned in the same exchange (IKE_AUTH). It's up to the responder to narrow the traffic selectors appropriately (for TSi to only the virtual IP, or if it has additional remote subnets configured to those too).

If the peer supports childless IKE_SAs (there should be a notify in the IKE_SA_INIT response), you could try setting `connections.<conn>.childless` to `force` (only possible via [swanctl.conf](https://docs.strongswan.org/docs/5.9/swanctl/swanctlConf.html)). Then the first CHILD_SA is initiated in a separate CREATE_CHILD_SA exchange instead of during IKE_AUTH and strongSwan will narrow `dynamic` to the virtual IP already in the proposed TSi.
